File: client/typing_sounds.py
"""Cross-platform typing-feedback cue definitions and resolution."""

from __future__ import annotations

from dataclasses import dataclass


TYPING_SOUND_FAMILY = "typing"
TYPING_SOUND_HANDLE = "client:typing-feedback"
TYPING_SOUND_VOLUME = 0.5

_DIGIT_NAMES = (
    "zero",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
)
DIGIT_SOUND_ASSETS = tuple(
    f"typing_digit_{name}.ogg" for name in _DIGIT_NAMES
)
TYPING_DELETE_ASSET = "typing_delete.ogg"
TYPING_RETURN_ASSET = "typing_return.ogg"
TYPING_EXACT_ASSETS = (
    *DIGIT_SOUND_ASSETS,
    TYPING_DELETE_ASSET,
    TYPING_RETURN_ASSET,
)

_DELETE_KEYS = frozenset({"Backspace", "Delete"})
_RETURN_KEYS = frozenset({"Enter", "NumpadEnter"})


@dataclass(frozen=True, slots=True)
class TypingSoundCue:
    """A single exact asset or dynamically discovered numbered family."""

    asset: str = ""
    family: str = ""

    def __post_init__(self) -> None:
        if bool(self.asset) == bool(self.family):
            raise ValueError("A typing sound cue must specify exactly one source")


GENERIC_TYPING_CUE = TypingSoundCue(family=TYPING_SOUND_FAMILY)


def resolve_typing_sound_cue(
    key: str,
    *,
    modified: bool = False,
    auto_repeat: bool = False,
) -> TypingSoundCue | None:
    """Resolve one physical text-control key press to its feedback cue.

    Key-repeat events are intentionally silent. Delete and Return remain useful
    with modifiers because common text editing commands use Ctrl/Command with
    those keys. Other modified keys are treated as shortcuts.
    """

    if auto_repeat:
        return None
    if key in _DELETE_KEYS:
        return TypingSoundCue(asset=TYPING_DELETE_ASSET)
    if key in _RETURN_KEYS:
        return TypingSoundCue(asset=TYPING_RETURN_ASSET)
    if modified:
        return None
    if len(key) == 1 and key.isdecimal() and key.isascii():
        return TypingSoundCue(asset=DIGIT_SOUND_ASSETS[int(key)])
    if len(key) == 1 and key.isprintable():
        return GENERIC_TYPING_CUE
    return None

File: web_client/typing_sounds.js
export const TYPING_SOUND_FAMILY = "typing";
export const TYPING_SOUND_HANDLE = "client:typing-feedback";
export const TYPING_SOUND_VOLUME = 50;

const DIGIT_NAMES = Object.freeze([
  "zero",
  "one",
  "two",
  "three",
  "four",
  "five",
  "six",
  "seven",
  "eight",
  "nine",
]);

export const DIGIT_SOUND_ASSETS = Object.freeze(
  DIGIT_NAMES.map((name) => `typing_digit_${name}.ogg`),
);
export const TYPING_DELETE_ASSET = "typing_delete.ogg";
export const TYPING_RETURN_ASSET = "typing_return.ogg";
export const TYPING_EXACT_ASSETS = Object.freeze([
  ...DIGIT_SOUND_ASSETS,
  TYPING_DELETE_ASSET,
  TYPING_RETURN_ASSET,
]);

const DELETE_KEYS = new Set(["Backspace", "Delete"]);
const RETURN_KEYS = new Set(["Enter", "NumpadEnter"]);
const GENERIC_PHYSICAL_CODES = new Set([
  "Space",
  "Backquote",
  "Minus",
  "Equal",
  "BracketLeft",
  "BracketRight",
  "Backslash",
  "Semicolon",
  "Quote",
  "Comma",
  "Period",
  "Slash",
  "IntlBackslash",
  "IntlRo",
  "IntlYen",
  "NumpadDecimal",
  "NumpadAdd",
  "NumpadSubtract",
  "NumpadMultiply",
  "NumpadDivide",
]);

function digitFromEvent(key, code) {
  if (/^[0-9]$/.test(key)) {
    return Number(key);
  }
  const match = /^(?:Digit|Numpad)([0-9])$/.exec(code);
  return match ? Number(match[1]) : null;
}

function isPhysicalTextCode(code) {
  return /^Key[A-Z]$/.test(code) || GENERIC_PHYSICAL_CODES.has(code);
}

export function isImeCompositionKeyEvent(event) {
  return Boolean(event?.isComposing || event?.keyCode === 229);
}

export function resolveTypingSoundCue(event) {
  if (!event || event.repeat === true) {
    return null;
  }

  const key = String(event.key || "");
  const code = String(event.code || "");
  if (DELETE_KEYS.has(key) || code === "Backspace" || code === "Delete") {
    return { asset: TYPING_DELETE_ASSET };
  }
  if (RETURN_KEYS.has(key) || RETURN_KEYS.has(code)) {
    return { asset: TYPING_RETURN_ASSET };
  }
  const altGraph = event.getModifierState?.("AltGraph") === true;
  if (event.metaKey || ((event.ctrlKey || event.altKey) && !altGraph)) {
    return null;
  }

  const digit = digitFromEvent(key, code);
  if (digit !== null) {
    return { asset: DIGIT_SOUND_ASSETS[digit] };
  }
  if (
    (key.length === 1 && key.codePointAt(0) >= 32)
    || isPhysicalTextCode(code)
  ) {
    return { family: TYPING_SOUND_FAMILY };
  }
  return null;
}
